controller/share_manager_controller.go

```diff
@@ -439,30 +450,44 @@ func (c *ShareManagerController) detachShareManagerVolume(sm *longhorn.ShareMana
 // starting, running, error -> stopped (no longer required, volume detachment)
 // controls transitions to starting, stopped
 func (c *ShareManagerController) syncShareManagerVolume(sm *longhorn.ShareManager) (err error) {
-	var isNotNeeded bool
-	defer func() {
-		// ensure volume gets detached if share manager needs to be cleaned up and hasn't stopped yet
-		// we need the isNotNeeded var so we don't accidentally detach manually attached volumes,
-		// while the share manager is no longer running (only run cleanup once)
-		if isNotNeeded && sm.Status.State != longhorn.ShareManagerStateStopping && sm.Status.State != longhorn.ShareManagerStateStopped {
-			getLoggerForShareManager(c.logger, sm).Info("stopping share manager")
-			if err = c.detachShareManagerVolume(sm); err == nil {
+	log := getLoggerForShareManager(c.logger, sm)
+	volume, err := c.ds.GetVolume(sm.Name)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			if sm.Status.State != longhorn.ShareManagerStateStopped {
+				getLoggerForShareManager(c.logger, sm).Infof("stopping share manager because the volume %v is not found", sm.Name)
 				sm.Status.State = longhorn.ShareManagerStateStopping
 			}
+			return nil
 		}
-	}()
-
-	log := getLoggerForShareManager(c.logger, sm)
-	volume, err := c.ds.GetVolume(sm.Name)
-	if err != nil && !apierrors.IsNotFound(err) {
 		log.WithError(err).Error("failed to retrieve volume for share manager from datastore")
 		return err
 	}
 
-	if !c.isShareManagerRequiredForVolume(volume) {
-		if sm.Status.State != longhorn.ShareManagerStateStopping && sm.Status.State != longhorn.ShareManagerStateStopped {
-			log.Info("share manager is no longer required")
-			isNotNeeded = true
+	vaName := types.GetLHVolumeAttachmentNameFromVolumeName(volume.Name)
+	va, err := c.ds.GetLHVolumeAttachment(vaName)
+	if err != nil {
+		return err
+	}
```

Comment:
Can we simplify this to one function call like `GetLHVolumeAttachmentByVolume`?

Reply:
Thank you. Fixed